assets/js/modules/tagmanager/datastore/existing-tag.js

```diff
@@ -27,15 +27,29 @@ import invariant from 'invariant';
 import API from 'googlesitekit-api';
 import Data from 'googlesitekit-data';
 import { STORE_NAME } from './constants';
-import { getExistingTag } from '../../../util/tag';
+import { STORE_NAME as CORE_SITE } from '../../../googlesitekit/datastore/site/constants';
+import { getExistingTagURLs, extractExistingTag } from '../../../util/tag';
+import { tagMatchers } from '../util';
 import { isValidContainerID, isValidContainerSelection } from '../util/validation';
 import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
 
 const { createRegistrySelector } = Data;
 
 const fetchGetExistingTagStore = createFetchStore( {
 	baseName: 'getExistingTag',
-	controlCallback: () => getExistingTag( 'tagmanager' ),
+	controlCallback: async () => {
+		const existingTagURLs = await getExistingTagURLs( Data.select( CORE_SITE ) );
+		let tagFound = null;
+		for ( const url of existingTagURLs ) {
+			const html = await Data.select( CORE_SITE ).getHTMLForURL( url );
+			// TODO: html ends up being undefined here.
```

Comment:
@aaemnnosttv From what I can tell, my main issue with the tests arises here, as this `html` ends up being `undefined`. I think the `getHTMLForURL()` selector initially returns `undefined`, since it doesn't have a value, so then it makes the fetch request to get the html for the given url.  So I think I need to wait until the selector has done all that, but I'm not sure how to go about that here within this `controlCallback`.

Also, is there a way to do `createRegistryControl` with these `controlCallbacks`? If I try to use it for the callback I get errors, which is why I'm using `Data.select` instead of the `registry.select`.

Reply:
> From what I can tell, my main issue with the tests arises here, as this `html` ends up being `undefined`. I think the `getHTMLForURL()` selector initially returns `undefined`, since it doesn't have a value, so then it makes the fetch request to get the html for the given url.  So I think I need to wait until the selector has done all that, but I'm not sure how to go about that here within this `controlCallback`.

If the selector uses a resolver to fulfill the value then it will return whatever the current value is - in this case `undefined` makes sense. Since selectors are synchronous you can't wait for anything in a selector. However, components that use this selector will update when the datastore updates (i.e. receives the HTML for a URL) so when they select it again it will use whatever the current state is at that time.

For this reason we generally support `undefined` as an expected value which always means that the data hasn't been loaded yet.

> Also, is there a way to do `createRegistryControl` with these `controlCallbacks`? If I try to use it for the callback I get errors, which is why I'm using `Data.select` instead of the `registry.select`.

Unfortunately not. That is something which was intentional in the design. If you do need a registry control however, you don't _need_ to use `createFetchStore`. It's possible we may update this to support using a registry control in the future if we have sufficient need for it but so far we haven't.

One reason why you may be having problems with tests may be from using `Data` directly. You can't use `Data` directly, especially in a datastore as we use many different registry instances in tests, none of which are instance returned by the `googlesitekit-data` import. `Data` should only be referenced directly for things like `useSelect`  or `combineStores`. The only exception to this are entrypoint files which are only really loaded in the browser/production.